IMOD/manpages/realscalemod.man
.na
.nh
.TH realscalemod 1 4.6.34 IMOD
.SH NAME
realscalemod - to scale a model to microns, adjusting for tilts
.SH SYNOPSIS
realscalemod
.SH DESCRIPTION
REALSCALEMOD will take a model and rescale it into microns.
It will adjust not only for pixel size and section thickness but
also for any tilts during microscopy.
.P
If the sections were significantly tilted during microscopy, the
program can adjust for these tilts given the proper information.
Prepare a file in which the first line shows the Z value and the
tilt of the first tilted section (or of the first section, if that
one was tilted), and each successive line shows the Z value and tilt
for each section on which tilt was changed.  Z values should occur in
ascending order.
.P
Entries to the program are:
.P
Name of model file to be scaled
.P
Name of output file in which to place scaled model
.P
Name of file with tilt information, or Return if none
.P
IF the model has no scaling information in its header, also make the
following three entries:
.P
     Magnification of negatives (without any commas)
.P
     Scale, in microns per pixel, at which negatives were digitized
.P
     Section thickness in nanometers
.P
.SH HISTORY
.nf
  Written by David Mastronarde, 11/19/93; modified for IMOD 4/24/97
.fi
.SH BUGS
Email bug reports to mast@colorado.edu.


IMOD/History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/23: Fixed crash in Ctfphaseflip with X size of 4092 (i.e., K3 camera) and
tilt angle of exactly 0; fixed strip width at exactly 0 tilt with a maximum
width entered (In 4.11)

9/22/23: In Etomo - Fidual Model, fixed the "Use Adjusted Track Com File"
button, which was not functioning (In 4.11).(Bug# 2458)

3/14/23: Fixed Batchruntomo problem running Splitcorrection with Python 3
(In 4.11).

2/3/23: IMOD_4-11-24
2/3/23: Fixed Autofidseed problem with Python 3 (In 4.11).

1/10/22: IMOD_4-11-23
1/10/22: Fixed Sampletilt when the dataset name contains "slice" (In 4.11).

1/8/22: IMOD_4-11-22
12/22/22: Fixed a problem where most of the interfaces are unable to handle a
missing file name style (In 4.11).(Bug# 2427)

12/21/22: Fixed initialization of the back projection checkbox in batch
interface (in 4.11).(Bug# 2417)

12/7/22: Fixed Tiltalign crash from array that was too small when there are
many views and few points (In 4.11, already fixed in 4.12).

11/1/22: IMOD_4-11-21
11/1/22: Fixed failure in Copytomocoms from not converting some directive
values to float (In 4.11).

9/27/22: Add entry for autofidseed LowerTargetForClustered option.(Bug# 2418)

9/27/22: In Etomo - Batchruntomo Interface, fixed a problem with the
backprojection checkbox (In 4.11).(Bug# 2417)

9/7/22: Changes to framewatcher to watch for EER files (In 4.11).

9/2/22: IMOD_4-11-20
9/1/22: In Etomo - Batchruntomo Interface, fixed some problems with the back-
projection checkbox (In 4.11).(Bug# 2417)

8/25/22: Switched to Qt 5.15.12 for Linux builds to solve a problem with being
unable to resize or move windows after mouse-rotating in slicer (In 4.11).

8/3/22: Fixed 3dmod crash if model loaded during initial image load (In 4.11).

7/29/22: IMOD_4-11-19
7/29/22: Fixed no new line before MESSAGE: from Tilt if GPU not available
(In 4.11).

7/28/22: In Etomo - PEET, fixed a problem where the processchunks monitor did
not recognize any single-line messages from processchunks.out (In 4.11).
(Bug# 2321)

7/18/22: Changed defocus tolerance in ctfcorrection command file to 50 and
made Ctf3dsetup and Subtomosetup use the slab thickness for this tolerance, 
since it determines the width of regions near the edges corrected at just one
defocus (In 4.11).

7/12/22 In Etomo - Final Aligned Stack handling a config file with an
incorrect path (In 4.11).(Bug# 2404)

7/8/22 In Etomo - Tomogram Generation fixed the choosing of SIRT output
tomograms to load into 3dmod (In 4.11).(Bug# 2402)

7/7/22 In Etomo - Reconstruction Interface added missing image file name style
recovery (In 4.11).(Bug# 2403)

7/1/22: Re-enabled use of RGB data in Drawing Tools (In 4.11).

6/30/22: In Etomo - Batchruntomo Interface fixed a problem where a delivery
was done on every run (In 4.11).((Bug# 2405)

6/22/22: The minimum zero spacing in CTF correction is now defined as a float
and setting it will no longer generate an error in Etomo (In 4.11).(Bug# 2406)

6/18/22: IMOD_4-11-18
6/17/22: Fixed summing into power spectrum in Ctfplotter, which was off by a
pixel in the upper half of the FFT and biased defocus by 10-20 nm (In 4.11).

6/15/22: IMOD_4-11-17
6/14/22: Fixed wrong scaling of spherical aberration in Ctfphaseflip, which
made it correct effectively 20-50 nm higher defocus and get off above ~5 A
resolution (In 4.11).

6/11/22: Fixed Ctfplotter with pixel cropping to take tiles from correct
locations; it was sampling more from one side than the other, not adjusting
each spectrum properly for tile location, blurring the power spectrum, and
giving increasingly biased defocus estimates with increasing tilt (In 4.11).

6/10/22: In Etomo - Reconstruction interface, fixing a problem that prevented
a dataset created by batchruntomo from being loaded (In 4.11).(Bug# 2409)

6/8/22: Ctfplotter: fixed the crop pixel option when running
non-interactively; it was either not cropping or going into an infinite
loop. Fixed the starting point for autofitting to be near zero tilt 
when running non-interactively.  Fixed a memory leak (In 4.11).

6/6/22: Fixed Zap and Slicer snapshots having rubberband on edge with high-DPI
monitors (In 4.11).

6/2/22: Fixed crash in Ctfphaseflip with GPU from accessing strip coordinates
out of range (In 4.11).

5/30/22: IMOD_4-11-16
5/30/22: Made sure autocontrasting does not happen until the Zap window has
been drawn (In 4.11).

5/26/22: Made Batchruntomo detect and recover from an Etomo error in dataset
location when restarting from a step (In 4.11).

5/26/22: Fixed tile cache display in Zap window sometimes not being refreshed
after all the tiles are loaded (In 4.11).

5/25/22: Made Makepyramid work with a single-image file (In 4.11).

5/22/22: Changed installer for Linux/Mac to put link to python, if needed, in
a separate directory that is put on the back of the path (In 4.11).

5/20/22: Fixed skipping of global clip planes in model view (In 4.11).

5/20/22: Fixed sychronization of ghost and model view image display checkboxes
when toggling with hot keys (In 4.11).

5/20/22: Stopping limiting EER Z-summing to 99 in 3dmod (In 4.11).

5/6/22: Fixed a test for extension types in Splitblend, Splittilt, and
Splitcorrection (in 4.11).

4/26/22: Fixed direction of Step Up/Down buttons in Ctfplotter, broken 
1/22/20, and improved treatment of autofitting range (In 4.11).

4/19/22: Stopped ctfplotter from interpreting a zero astigmatism as not being
measured at all (In 4.11).

3/29/22: IMOD_4-11-15
3/29/22: Fixed Alignframes making unweighted sum with CPU (In 4.11).

3/21/22: IMOD_4-11-14
3/21/22: Changed installer for Linux to make link to python3 or python2 in
IMOD/bin instead of in /usr/bin when there is no "python" on the path, but for
Mac OS 12 it will look for python3 from python.org, homebrew, or Apple Developer
Tools and make the link directly to the first one it finds (In 4.11). 


IMOD/flib/model/realscalemod.f
*       * * * * REALSCALEMOD * * * * *
c       
c       REALSCALEMOD will take a model and rescale it into microns.
c       It will adjust not only for pixel size and section thickness but
c       also for any tilts during microscopy.
c       
c       If the sections were significantly tilted during microscopy, the
c       program can adjust for these tilts given the proper information.
c       Prepare a file in which the first line shows the Z value and the
c       tilt of the first tilted section (or of the first section, if that
c       one was tilted), and each successive line shows the Z value and tilt
c       for each section on which tilt was changed.  Z values should occur in
c       ascending order.
c       
c       Entries to the program are:
c       
c       Name of model file to be scaled
c       
c       Name of output file in which to place scaled model
c       
c       Name of file with tilt information, or Return if none
c       
c       IF the model has no scaling information in its header, also make the
c       following three entries:
c       
c       .  Magnification of negatives (without any commas)
c       
c       .  Scale, in microns per pixel, at which negatives were digitized
c       
c       .  Section thickness in nanometers
c       
c       David Mastronarde  11/19/93; modified for IMOD 4/24/97
c       
      use fortmodel
      character*320 modelfile,newmodel,tiltfile
      logical exist,readw_or_imod
      real*4 tiltzstrt(1000),tilt(1000),costilt(1000),remapz(1000)
c       
      call setExitPrefix('ERROR: REALSCALEMOD - ')
91    write(*,'(1x,a,$)')'Name of input model file: '
      read(*,'(a)')modelfile
c       
c       read in the model
c       
75    exist=readw_or_imod(modelfile)
      if(.not.exist)go to 91
c       
      write(*,'(1x,a,$)')'Name of output file for scaled model: '
      read(*,'(a)')newmodel
c       
      write(*,'(1x,a,$)')
     &    'Name of file of tilt info (Return if none): '
      read(*,'(a)')tiltfile
      if(tiltfile.ne.' ')then
        call dopen(3,tiltfile,'ro','f')
        ntilts=0
3       i=ntilts+1
        read(3,*,end=5)tiltzstrt(i),tilt(i)
        ntilts=i
        costilt(ntilts)=cosd(tilt(ntilts))
        go to 3
5       remapz(1)=tiltzstrt(1)
        do i=1,ntilts-1
          remapz(i+1)=remapz(i)+(tiltzstrt(i+1)-tiltzstrt(i))
     &        /costilt(i)
        enddo
        zmax=-1.e10
        do i=1,n_point
          zz=p_coord(3,i)
          if(zz.ge.tiltzstrt(1))then
            itilt=ntilts
            do while(zz.lt.tiltzstrt(itilt))
              itilt=itilt-1
            enddo
            znew=remapz(itilt)+(zz-tiltzstrt(itilt))/
     &          costilt(itilt)
            p_coord(3,i)=znew
            zmax=max(zmax,znew)
          endif
        enddo
        print *,'new maximum z is',zmax
      endif
c       
      defscal=1.e6
      ierr=getimodhead(xyscal,zscale,xofs,yofs,zofs,ifflip)
      if(ierr.eq.0.and.abs(xyscal-defscal)/defscal.gt.1.e-5)then
        write(*,'(a,f10.6,a)')' Scale set from model header at',
     &      xyscal,' microns/pixel'
        zscal=xyscal*zscale
      else
        ifflip=0
        write(*,'(1x,a,$)')'Magnification of negatives: '
        read(*,*)xmag
        write(*,'(1x,a,$)')'Scale at which negatives were digitized'
     &      //' (microns per pixel from VIDS): '
        read(*,*)umperpix
        write(*,'(1x,a,$)')'Nominal section thickness (nm): '
        read(*,*)secthick
c         
        xyscal=umperpix/xmag
        zscal=secthick/1000.
      endif
      if(ifflip.eq.0)then
        indy=2
        indz=3
      else
        indz=2
        indy=3
      endif
c       
      do i=1,n_point
        p_coord(1,i)=p_coord(1,i)*xyscal
        p_coord(indy,i)=p_coord(indy,i)*xyscal
        p_coord(indz,i)=p_coord(indz,i)*zscal
      enddo
c       
77    call write_wmod(newmodel)
      print *, 'RESCALED MODEL WRITTEN'
      call exit(0)
      end
